mac/Keyman4MacIM/Keyman4MacIM/KMContext.m

```diff
@@ -0,0 +1,201 @@
+ * Also useful for applications that do not allow us to read the context.
+ */
+
+#import "KMContext.h"
+#include <Carbon/Carbon.h> /* For kVK_ constants. */
+
+NSUInteger const MAXIMUM_CONTEXT_LENGTH = 80;         // KM_KBP_IT_CHAR
+
+@interface KMContext()
+
+@property (readonly) NSMutableString *context;
+@property BOOL invalid;
+
+@end
+
+@implementation KMContext
+
+-(instancetype)init {
+  self = [super init];
+  if (self) {
+    _context = [[NSMutableString alloc] initWithCapacity:MAXIMUM_CONTEXT_LENGTH];
+    _invalid = YES;
+  }
+  return self;
+}
+
+-(instancetype)initWithString:(NSString*)initialContext {
+  self = [self init];
+  if (self) {
+    [_context setString:[KMContext trimToMaximumContext:initialContext]];
+    _invalid = NO;
+  }
+  return self;
+}
+
++(NSString*)trimToMaximumContext:(NSString*)initialContext {
+  NSString *maximumContext = nil;
+
+  // copy up to MAXIMUM_CONTEXT_LENGTH from the end of the context
+  NSUInteger initialLength = [initialContext length];
+  if (initialLength > MAXIMUM_CONTEXT_LENGTH) {
+    maximumContext = [initialContext substringFromIndex:initialLength-MAXIMUM_CONTEXT_LENGTH];
+  } else {
+    maximumContext = initialContext;
+  }
+
+  return maximumContext;
+}
+
+-(void)resetContext:(NSString*)newContext {
+  [_context setString:[KMContext trimToMaximumContext:newContext]];
+  self.invalid = NO;
+  NSLog(@"   ***CachedContext resetContext, resulting context = %@", _context);
+}
+
+-(NSString*)currentContext {
+  return (NSString*)_context;
+}
+
+-(BOOL)isInvalid {
+  return self.invalid;
+}
+
+-(void)invalidateContext {
+  [self clearContext];
+  self.invalid = YES;
+}
+
+-(void)clearContext {
+  _context.string = @"";
+  NSLog(@"   ***CachedContext clearContext, resulting context = %@", _context);
+}
+
+-(void)appendMarker {
+  #define UC_NONCHARACTER             0xFFFF
+  NSString *markerString = [NSString stringWithFormat:@"%C", UC_NONCHARACTER];
+  [self.context appendString:markerString];
+  NSLog(@"   ***CachedContext appendMarker, resulting context = %@", _context);
+}
+
+/**
+ * Deletes the last complete code point in the context string, if it is not empty.
+ * NSString is UTF-16, so one code point could be composed of multiple code units.
+ * Checks for the size of the final code point to see how much needs to be deleted.
+ */
+-(void)deleteLastCodePoint {
+  NSUInteger contextLength = self.context.length;
+  
+  if (contextLength > 0) {
+    NSUInteger characterIndex = contextLength-1;
+    NSRange characterRange = [self.context rangeOfComposedCharacterSequenceAtIndex:characterIndex];
+    NSLog(@"KMContext deleteLastUnicodeCharacter, index = %lu, rangeOfLastCharacter = %@\n", characterIndex, NSStringFromRange(characterRange));
+    [self.context deleteCharactersInRange:characterRange];
+  }
+  NSLog(@"   ***CachedContext deleteLastCodePoint, resulting context = %@", _context);
+}
+
+-(void)replaceSubstring:(NSString*)newText count:(int)count {
```

Comment:
Is `count` number of codepoints or number of code units? We should document that